runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/control/DefaultJobBundleFactoryTest.java

```diff
@@ -360,6 +360,19 @@ public void closesEnvironmentOnCleanup() throws Exception {
     verify(remoteEnvironment).close();
   }
 
+  @Test
+  public void closesEnvironmentOnCleanupWithPendingRefs() throws Exception {
+    try (DefaultJobBundleFactory bundleFactory =
+        createDefaultJobBundleFactory(envFactoryProviderMap)) {
+      DefaultJobBundleFactory.SimpleStageBundleFactory stageBundleFactory =
+          (DefaultJobBundleFactory.SimpleStageBundleFactory)
+              bundleFactory.forStage(getExecutableStage(environment));
+      // The client is still being used, e.g. when the pipeline fails and is shut down
+      stageBundleFactory.currentClient.wrappedClient.ref();
```

Comment:
What would cause this extra ref in the actual operator lifecycle?

Reply:
I'm assuming here we have a `ref()` due to the client being used. If we had no ref, we would be idling. The error message we were seeing is only possible if we are currently processing a bundle.